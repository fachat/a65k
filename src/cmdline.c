
/****************************************************************************

    commandline parsing
    Copyright (C) 2016-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdlib.h>
#include <strings.h>
#include <string.h>

#include "hashmap.h"
#include "array_list.h"
#include "cmdline.h"
#include "err.h"
#include "infiles.h"

// hash from param name to cmdline_t sruct for quick find
static hash_t *params = NULL;
// list of parameters for help output
static list_t *paramlist = NULL;

static const char *prg_name = NULL;

static err_t usage(int flag, void* extra) {
	printf("Usage: %s [options] in-filename1 [in-filename2 ...]\n"
		"Cross-assembler for 65xx/R65C02/65816/65CE02/65002\n"
		"\n", prg_name); 

	list_iterator_t *iter = list_iterator(paramlist);
	cmdline_t *param = NULL;
	while ( (param = list_iterator_next(iter)) ) {
		printf("  -%s\n\t%s\n", param->name, param->description);
	}
	
	return E_ABORT;
}

static const cmdline_t help[] = {
	{ "?", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
	{ "help", PARTYPE_FLAG, NULL, usage, NULL, "Show this help" },
};

static const char *key_from_param(const void *entry) {
	return ((cmdline_t*)entry)->name;
}

// TODO init the cmdline parser with the name as which the program was called
// for example "a65k", but also "xa65"
void cmdline_module_init() {

	params = hash_init_stringkey(50, 25, key_from_param);
	paramlist = array_list_init(20);

	cmdline_register_mult(help, sizeof(help)/sizeof(cmdline_t));
}

// template method where extra_param is the pointer to an int variable to set
err_t cmdline_set_flag(int flag, void *extra_param) {
	*((int*)extra_param) = flag;
	return E_OK;
}

void cmdline_register_mult(const cmdline_t *param, int num ) {
	for (int i = 0; i < num; i++) {
		cmdline_register(&param[i]);
	}
}

void cmdline_register(const cmdline_t *param) {
	
	if (hash_put(params, (void*)param) != NULL) {
		// must not happen
		exit(1);
	}
	list_add(paramlist, (void*)param);
}

err_t cmdline_parse(int argc, char *argv[]) {

	err_t rv = E_OK;

	prg_name = argv[0];

        int i = 1;
        while (i < argc && !rv) {
                if (argv[i][0] == '-') {
			char *val = NULL;
			int flag = 1;
			const char *name = argv[i]+1;
			if (name == strstr(name, "no-")) {
				name = name+3;
				flag = 0;
			}
			char *end = index(name, '=');
			if (end != NULL) {
				val = end + 1;
				end[0] = 0;
			}
			
			// lookup option
			cmdline_t *opt = hash_get(params, name);
			if (opt != NULL) {
				switch (opt->type) {
				case PARTYPE_FLAG:
					rv = opt->setflag(flag, opt->extra_param);
					break;
				case PARTYPE_PARAM:
					// TODO: error if param is missing
					// TODO: if started with "no-" return error
					rv = opt->setfunc(val, opt->extra_param);
					break;
				}
			} else {
				// TODO error cmdline parameter not found
			}
                } else {
			// register file with snapshot (clone) of current parser configuration
                        infiles_register(argv[i], parser_current_config());
                }   
		i++;
        }   
	return rv;
}




