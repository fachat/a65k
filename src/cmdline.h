
/****************************************************************************

    commandline parsing
    Copyright (C) 2016 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef CMDLINE_H
#define CMDLINE_H

#include "err.h"

// init the cmdline parser 
void cmdline_module_init();

// parse the options
err_t cmdline_parse(int argc, char *argv[]);

typedef enum {
	// no parameter, sets a bool when present with/without "no-" prefix
	PARTYPE_FLAG,
	// with required parameter
	PARTYPE_PARAM
} param_type_t;

typedef struct {
	// name of cmdline param
	const char 	*name;		

	// option has a parameter
	param_type_t	type;

	// function to call when parsed and type is PARTYPE_PARAM
	err_t 		(*setfunc)(const char *value, void *extra_param);

	// function to call when parsed and type is PARTYPE_FLAG
	void 		(*setflag)(int flag, void *extra_param);

	// extra param to be passed to set function
	void		*extra_param;

	// description
	const char	*description;
} cmdline_t;

// template method where extra_param is the pointer to an int variable to set
void cmdline_set_flag(int flag, void *extra_param);
 
void cmdline_register(const cmdline_t *param);

#endif

