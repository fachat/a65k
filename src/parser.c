/****************************************************************************

    parser
    Copyright (C) 2015-2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#define	DEBUG

#include <stdio.h>
#include <ctype.h>

#include "err.h"

#include "log.h"
#include "mem.h"
#include "list.h"
#include "array_list.h"
#include "hashmap.h"
#include "position.h"

#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "parser.h"
#include "label.h"
#include "block.h"
#include "tokenizer.h"
#include "arith.h"
#include "parser-config.h"

typedef struct {
	block_t			*blk;

	list_t			*statements;
} parser_t;


typedef enum {
	P_INIT,		// start of line, accept label definitions, operations and pseudos
	P_OP,		// has parsed label, accept "=" or operation
	P_PARAM,	// after operation, parse parameters to the operation
	P_COLON,	// after either label or line number, where a colon can be
	P_END,		// expect end of line
} pstate_t;


static const type_t statement_memtype = {
	"statement_t",
	sizeof(statement_t)
};

static const type_t parser_memtype = {
	"parser_t",
	sizeof(parser_t)
};

static parser_t *p = NULL;

void parser_module_init(void) {

	p = mem_alloc(&parser_memtype);
	p->blk = NULL;
	p->statements = array_list_init(10000);

	parser_config_init();
}


static statement_t *new_statement(const context_t *ctx) {
	statement_t *stmt = mem_alloc(&statement_memtype);

	stmt->type = S_NONE;
	stmt->blk = p->blk;
	stmt->ctx = ctx;
	stmt->op = NULL;
	stmt->label = NULL;
	stmt->syn = SY_IMP;

	// negative line number is not set
	stmt->lineno = -1;

	stmt->rs_prefix = RS_NOT_SET;
	stmt->le_prefix = LE_NOT_SET;
	stmt->um_prefix = 0;
	stmt->nf_prefix = 0;

	stmt->comment = NULL;

	return stmt;
}

static statement_t *new_statement_in_line(const context_t *ctx, statement_t *prev) {

	statement_t *stmt = new_statement(ctx);

	stmt->lineno = prev->lineno;

	return stmt;
}


static void statement_push(const context_t *ctx, statement_t *stmt) {

	// statements in parse order
        list_add(p->statements, stmt);

	// statements per segment/region
	segment_push_statement(ctx->segment, stmt);
}


void parser_reset() {
        // only for tests - ignore that we would have to mfree stuff before
        p->statements = array_list_init(10000);
}

static inline void error_double_prefix(const tokenizer_t *tok, char prefix, char conflict) {
        toklog_error(tok, "Double prefix '%c' conflicting with previous '%c'", prefix, conflict);
}

static inline void warn_double_prefix(const tokenizer_t *tok, char prefix) {
        toklog_warn(tok, "Double prefix '%c'", prefix);
}

static inline void warn_operation_not_for_cpu(const tokenizer_t *tok, const char *op_name, const char *cpu_name) {
        toklog_warn(tok, "Operation name %s is not available for CPU %s, assuming label!", op_name, cpu_name);
}

static inline void error_assign_without_label(const tokenizer_t *tok) {
	toklog_error(tok, "%s", "Assign without label");
}


/**
 * parse the operation parameter incl. addressing mode
 */
err_t parse_param(tokenizer_t *tok, statement_t *stmt) {

	if (tok->type == T_TOKEN && tok->vals.op == OP_HASH) {

		stmt->syn = SY_IMM;

		if (tokenizer_next(tok, 0)) {
			return arith_parse_integer(tok, stmt->blk, 0, (const ilist_t**)&stmt->param);	
		}
		return E_OK;
	} else {
		err_t rv = arith_parse_integer(tok, stmt->blk, 1, (const ilist_t**)&stmt->param);	
		if (rv == E_OK && stmt->param->len > 0) {

			stmt->syn = SY_ABS;

			anode_t *outer = NULL;
			anode_t *inner = NULL;
			anode_t *base = NULL;
			asub_type brtype = 0;
			anode_t *last = NULL;

			// unwrap brackets and/or index tokens (',x')

			// 1st unwrap ending indexes, just ",y", ",z"  
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX && (last->op == OP_YIND || last->op == OP_ZIND)) {
				outer = ilist_pop(stmt->param);
			}

			// 2nd unwrap brackets
			anode_t *first = ilist_get(stmt->param, 0);
			if (first->type == A_BRACKET && ilist_len(stmt->param) == 1) {
				// outer node is a bracket, and it is the only one
				brtype = first->val.subv.type; 
				ilist_t *basenode = first->val.subv.value;
				stmt->param = basenode;

				anode_t *first = ilist_get(stmt->param, 0);
				if (brtype == AB_RND 
						&& first->type == A_BRACKET 
						&& first->val.subv.type == AB_RND 
						&& ilist_len(stmt->param) == 1) {
					brtype = AB_DBLRND;
					ilist_t *basenode = first->val.subv.value;
					stmt->param = basenode;
				}
			}

			// 3rd from resulting unwrap ,x index
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX && last->op == OP_XIND) {
				inner = ilist_pop(stmt->param);
			}

			// 4th finally unwrapp base (,s / ,b / ,p) 
			last = ilist_last(stmt->param);
			if (last && last->type == A_INDEX) {
				base = ilist_pop(stmt->param);
				stmt->base = base->op;
			}

			// then compute syntax identifier
			// note that base is independent of syntax
			if (inner != NULL) {
				if (outer != NULL) {
					rv = E_SYNTAX;
				}
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSX;
					break;
				case AB_RND:
					stmt->syn = SY_XIND;
					break;
				case AB_RCT:
					stmt->syn = SY_XINDL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_XINDQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_XINDD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
			} else if (outer != NULL && outer->op == OP_YIND) {
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSY;
					break;
				case AB_RND:
					stmt->syn = SY_INDY;
					break;
				case AB_RCT:
					stmt->syn = SY_INDYL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_INDYQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_INDYD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
			} else if (outer != NULL && outer->op == OP_ZIND) {
				switch (brtype) {
				case 0:
					stmt->syn = SY_ABSZ;
					break;
				case AB_RND:
					stmt->syn = SY_INDZ;
					break;
				case AB_RCT:
					stmt->syn = SY_INDZL;
					break;
				case AB_DBLRCT:
					stmt->syn = SY_INDZQ;
					break;
				case AB_DBLRND:
					stmt->syn = SY_INDZD;
					break;
				default:
					// borken
					rv = E_SYNTAX;
					break;
				}
				
			
			}
		}
		return rv;
	}
}

static err_t parse_prefix(tokenizer_t *tok, statement_t *stmt) {

	err_t rv = E_OK;

	for (int i = 0; i < tok->len && !rv; i++) {
		char c = tolower(tok->line[tok->ptr + i]);
		switch(c) {
		case 'u':
			if (stmt->um_prefix) {
				warn_double_prefix(tok, c);
			}
			stmt->um_prefix = 1;
			break;
		case 'n':
			if (stmt->nf_prefix) {
				warn_double_prefix(tok, c);
			}
			stmt->nf_prefix = 1;
			break;
		case 'b':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_BYTE;
			break;
		case 'w':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_WORD;
			break;
		case 'l':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_LONG;
			break;
		case 'q':
			if (stmt->rs_prefix != RS_NOT_SET) {
				error_double_prefix(tok, c, prefix_rs_char(stmt->rs_prefix));
				rv = E_SYNTAX;
			}
			stmt->rs_prefix = RS_QUAD;
			break;
		case 'e':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_E;
			break;
		case 's':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_S;
			break;
		case '0':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_0;
			break;
		case '1':
			if (stmt->le_prefix != LE_NOT_SET) {
				error_double_prefix(tok, c, prefix_le_char(stmt->le_prefix));
				rv = E_SYNTAX;
			}
			stmt->le_prefix = LE_1;
			break;
		default:
			rv = E_SYNTAX;
			break;
		}
	}
	return rv;
}

err_t parser_push(const context_t *ctx, const line_t *line) {
	err_t rv = E_OK;

	if (line == NULL) {
		return E_END;
	}

	const parser_config_t *cfg = line->parsercfg;
	position_t *pos = line->position;

	// is the first block already set?
	if (p->blk == NULL) {
		p->blk = block_init(NULL, pos);
	}

	statement_t *stmt = new_statement(ctx);

	const instruction_t *op = NULL;
	const char *name = NULL;
	label_t *label = NULL;

	// tokenize the line
	tokenizer_t *tok = tokenizer_create(line->position, line->line, cfg->cstyle_allowed);

	tokenizer_next(tok, 0);

	while ((rv == E_OK) && tok->type != T_END && tok->is_valid) {

		while (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
			// ignore colon at beginning of line
			tokenizer_next(tok, 0);
		}

		// initial line number
		if (tok->type == T_LITERAL) {
			if (cfg->initial_lineno && (tok->vals.literal.type == LIT_DECIMAL)) {
				// line number parsing
				stmt->lineno = tok->vals.literal.value;

				tokenizer_next(tok, 0);
			} 

			if (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
				// ignore colon directly after line number
				tokenizer_next(tok, 0);
			}
		}

		while (stmt->op == NULL && tok->type == T_NAME) {
			name = mem_alloc_strn(tok->line + tok->ptr, tok->len);
			op = instruction_find(name);
			if (op != NULL) {
				// check if the operation is compatible with the current CPU
				if (0 == (ctx->cpu->isa & op->isa)) {
					// TODO: config for either no message or error
					warn_operation_not_for_cpu(tok, name, ctx->cpu->name);
					op = NULL;
				}
			}
			if (op == NULL) {
				// label
				// TODO: redefinition?
				label = label_init(ctx, name, pos);
				if (stmt->label) {
					// we already had a label
					stmt->type = S_LABEQPC;
					statement_push(ctx, stmt);
					stmt = new_statement_in_line(ctx, stmt);
				}
				stmt->label = label;
		
				tokenizer_next(tok, 0);
				if (tok->type == T_END) {
					stmt->type = S_LABEQPC;
				} else
				if (tok->type == T_TOKEN && tok->vals.op == OP_COLON) {
					// if we have a label with a colon, ignore colon
					tokenizer_next(tok, 0);
				}
			} else {
				// operation
				stmt->type = S_OPCODE;
				stmt->op = op;
				tokenizer_next(tok, 0);
				if (tok->type == T_TOKEN && tok->vals.op == OP_DOT) {
					// parse prefix
					if (tokenizer_next_prefix(tok) && (tok->type == T_NAME)) {
						rv = parse_prefix(tok, stmt);
					} else {
						error_syntax(tok, "Expecting prefix notation after '.'");
						rv = E_SYNTAX;
					}
					if (rv) {
						break;
					}
					tokenizer_next(tok, 0);
				}
	
				if (tok->type != T_TOKEN 
					|| (tok->vals.op != OP_SEMICOLON
						&& (tok->vals.op != OP_DOUBLESLASH || cfg->cstyle_allowed)
						&& (tok->vals.op != OP_COLON))) {
	
					// parse parameters
					rv = parse_param(tok, stmt);
	
					if (rv != E_OK) {
						break;
					}
				}
			}
		}

		if (!stmt->op) {
			if (tok->type == T_TOKEN && tok->vals.op == OP_DOT) {
				if (tokenizer_next(tok, 0)) {
					rv = parse_pseudo(tok, stmt);
					stmt->type = S_PSEUDO;
				} else {
					rv = E_SYNTAX;
				}
				if (rv) {
					break;
				}
			} else
			if (tok->type == T_TOKEN && tokenizer_op_details(tok->vals.op)->is_assign) {
				if (stmt->label == NULL && tok->vals.op != OP_ASSIGNMULT) {
					error_assign_without_label(tok);
					rv = E_SYNTAX;
					break;
				}
				if (stmt->label == NULL && tok->vals.op == OP_ASSIGNMULT) {
					stmt->type = S_PCDEF;
				} else {
					stmt->type = S_LABDEF;
				}
				stmt->assign = tok->vals.op;
				tokenizer_next(tok, 0);
				// parse parameters
				rv = parse_param(tok, stmt);
				if (rv) {
					break;
				}
			} else
			if (tok->type != T_END
				&& tok->vals.op != OP_SEMICOLON
				&& (tok->vals.op != OP_DOUBLESLASH || cfg->cstyle_allowed)
				&& (tok->vals.op != OP_COLON)) {
				rv = E_SYNTAX;
				error_syntax(tok, "Expecting opcode");
				break;
			}
		}

		if (tok->type == T_TOKEN) {
			if ((tok->vals.op == OP_SEMICOLON
				|| (tok->vals.op == OP_DOUBLESLASH && cfg->cstyle_allowed))) {

				if (tokenizer_next_comment(tok, cfg->colon_in_comments)) {
					stmt->comment = mem_alloc_strn(tok->line+tok->ptr, tok->len);

					tokenizer_next(tok, 0);
				}
			} 
		}

		if (tok->type == T_TOKEN) {
			if (tok->vals.op == OP_COLON) {
	
				statement_push(ctx, stmt);
				stmt = new_statement_in_line(ctx, stmt);

				tokenizer_next(tok, 0);
			} else {
				rv = E_SYNTAX;
				error_syntax(tok, "Expecting comment or colon");
			}
		}
	}

	if (rv == E_OK || rv == E_END) {
		statement_push(ctx, stmt);
	}
	tokenizer_free(tok);

	return rv;
}


list_iterator_t *parser_get_statements(void) {

	return list_iterator(p->statements);
}

