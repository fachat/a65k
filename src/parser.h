/****************************************************************************

    parser
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PARSER_H
#define PARSER_H

#include "context.h"
#include "label.h"
#include "block.h"
#include "operation.h"
#include "tokenizer.h"
#include "arith.h"
#include "list.h"

typedef enum {
        S_LABEQPC,              // set label to PC
        S_LABDEF,               // set label from parameter
	S_OPCODE,		// opcode
	S_PSEUDO,		// pseudo opcode
} stype_t;

/**
 * The statement_t struct is the main structure of the internal workings.
 * The parser creates a statement_t with all information to be gathered from the source.
 * The printer can "lint" this statement and pretty-format it
 * The assembler itself resolves labels and computes references and extends it with binary values
 * The printer can then also print the extended statement with binary values
 * The renderer can then put (a list of) statements into a binary format, depending on the output options
 */
typedef struct {
        const block_t           *blk;		// pointer to current block structure for label dereferencing
        const context_t         *ctx;		// current context, like CPU type, width, other options
        stype_t                 type;		// type of statement (see above)
        // optional
        const label_t           *label;		// TODO allow for multiple label definitions per line/statement
	// S_LABDEF
	const ilist_t		*setlabel;	// anode_t list for setting label to a value 
	// S_OPCODE
       	const operation_t       *op;		// pointer to info about operation, like "ADC", "TAX", ...
	op_t			base;		// base extension of the parameter, like ",B", ",s" if any, zero else
	
	syntax_type		syn;		// addressing mode syntax, see SY_* constants
       	ilist_t  	        *param;		// list of anode_t structs that describe the syntax tree for the parameter value
} statement_t;


void parser_module_init(void);

err_t parser_push(const context_t *context, const line_t *line);

list_iterator_t *parser_get_statements(void);

// for test cases only
void parser_reset();

#endif

