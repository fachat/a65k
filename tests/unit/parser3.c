

#include <stdio.h>

#include "types.h"
#include "position.h"
#include "cpu.h"
#include "segment.h"
#include "context.h"
#include "position.h"
#include "operation.h"
#include "cmdline.h"
#include "parser.h"
#include "print.h"
#include "log.h"


static void test(const context_t *ctx, position_t *pos, const char *txt) {

	parser_reset();

	line_t line = { txt, pos, parser_current_config() };

	do_print("");
	do_print(">>> %s", line.line);
		
	err_t err = parser_push(ctx, &line);

	if (err) {
		do_print("err=%d", err);
	}
	//log_set_level(LEV_DEBUG);

	list_iterator_t *iter = parser_get_statements();
	while (list_iterator_has_next(iter)) {
		statement_t *stmt = list_iterator_next(iter);
		print_debug_stmt(stmt);
	}
}


int main(int argc, char *argv[]) {

        (void)argc;
        (void)argv;

	cmdline_module_init();
	cpu_module_init();
	segment_module_init();
	operation_module_init();
	parser_module_init();
	tokenizer_module_init();

	//parser_config_init();

	char *pars[] =  { "a65k", "--parse-initial-lineno", "--no-parse-colon-in-comments"  };
	cmdline_parse(3, pars);

	const cpu_t *cpu = cpu_by_name("nmos");

	segment_t *seg = segment_new(NULL, "test", SEG_ANY, cpu->type, true);

	const context_t *ctx = context_init(seg, cpu);

	position_t pos = { "bogusfile", 1 };

	char *pars2[] =  { "a65k", "--parse-colon-in-comments" };
	cmdline_parse(2, pars2);

	test(ctx, &pos, "10020 LDA #$E0:STA Z0+1");	
}



