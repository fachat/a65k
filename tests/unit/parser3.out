WRN:
WRN:>>> 10020 LDA #$E0:STA Z0+1
WRN:--------------- STMT ---------------
WRN:LINE: 10020:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 1, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)224/e0
WRN:--------------- STMT ---------------
WRN:LINE: 10020:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:OPR: name:STA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 2, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=2
WRN:      type=l, modifier=0 ( ), op=0( ), label=Z0 (resolved=0)
WRN:      type=v, modifier=0 ( ), op=43(+), val=(d)1/1
WRN:
WRN:>>> 11010 SERI2 =$029B:SERI1 =$029C
WRN:--------------- STMT ---------------
WRN:LINE: 11010:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:SERI2 ctx:0 pos:bogusfile:1
WRN:LDEF: =
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)667/29b
WRN:--------------- STMT ---------------
WRN:LINE: 11010:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:SERI1 ctx:0 pos:bogusfile:1
WRN:LDEF: =
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)668/29c
WRN:
WRN:>>> 12100 *=$ECD9:BC .BYT 11,0
WRN:--------------- STMT ---------------
WRN:LINE: 12100:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:PCDEF: *=
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)60633/ecd9
WRN:--------------- STMT ---------------
WRN:LINE: 12100:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:BC ctx:0 pos:bogusfile:1
WRN:PSEUDO: byt
WRN:PARS:
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)11/b
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)0/0
WRN:
WRN:>>> 26040 PIEC
WRN:--------------- STMT ---------------
WRN:LINE: 26040:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:PIEC ctx:0 pos:bogusfile:1
WRN:LABEQPC: =*
WRN:
WRN:>>> 22350 .BYT <-1,KT4-KT,KT6-KT
WRN:--------------- STMT ---------------
WRN:LINE: 22350:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:PSEUDO: byt
WRN:PARS:
WRN:    A:len=1
WRN:      type=u, modifier=60(<), op=0( ), unary=45(-)
WRN:      A:len=1
WRN:        type=v, modifier=0 ( ), op=0( ), val=(d)1/1
WRN:    A:len=2
WRN:      type=l, modifier=0 ( ), op=0( ), label=KT4 (resolved=0)
WRN:      type=l, modifier=0 ( ), op=45(-), label=KT (resolved=0)
WRN:    A:len=2
WRN:      type=l, modifier=0 ( ), op=0( ), label=KT6 (resolved=0)
WRN:      type=l, modifier=0 ( ), op=45(-), label=KT (resolved=0)
WRN:
WRN:>>> 12000 .IF MIKE-1:.GOTO 12600
WRN:--------------- STMT ---------------
WRN:LINE: 12000:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:PSEUDO: if
WRN:PARS:
WRN:    A:len=2
WRN:      type=l, modifier=0 ( ), op=0( ), label=MIKE (resolved=0)
WRN:      type=v, modifier=0 ( ), op=45(-), val=(d)1/1
WRN:--------------- STMT ---------------
WRN:LINE: 12000:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:PSEUDO: goto
WRN:PARS:
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)12600/3138
WRN:
WRN:>>> 50000 .END; REM
WRN:err=3
WRN:--------------- STMT ---------------
WRN:LINE: 50000:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:PSEUDO: end
WRN:
WRN:>>> 28650 HC LDA #<-1:STA $C6
WRN:--------------- STMT ---------------
WRN:LINE: 28650:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:HC ctx:0 pos:bogusfile:1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 1, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=u, modifier=60(<), op=0( ), unary=45(-)
WRN:      A:len=1
WRN:        type=v, modifier=0 ( ), op=0( ), val=(d)1/1
WRN:--------------- STMT ---------------
WRN:LINE: 28650:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:OPR: name:STA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 2, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)198/c6
WRN:
WRN:>>> 28650 HC LDA #<--1:STA $C6
WRN:--------------- STMT ---------------
WRN:LINE: 28650:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:LAB: name:HC ctx:0 pos:bogusfile:1
WRN:OPR: name:LDA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 1, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=u, modifier=60(<), op=0( ), unary=45(-)
WRN:      A:len=1
WRN:        type=u, modifier=0( ), op=0( ), unary=45(-)
WRN:        A:len=1
WRN:          type=v, modifier=0 ( ), op=0( ), val=(d)1/1
WRN:--------------- STMT ---------------
WRN:LINE: 28650:
WRN:BLK:0 ctx:0 (CPU: nmos, isa:$7), seg=test
WRN:OPR: name:STA isa:$1561 isrel:0 acw:0 idxw:0
WRN:MOD: BASE: 0, SYN: 2, UM: 0, NF: 0, RS: 0, LE: 0
WRN:PAR: 
WRN:    A:len=1
WRN:      type=v, modifier=0 ( ), op=0( ), val=(h)198/c6
